### ISSUE TITLE
Clean up twisted.python.threadpool implementation using context managers

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5981]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5981|
|Type|enhancement|
|Created|2012-09-14 14:14:13Z|
|Branch|https://github.com/twisted/twisted/tree/threadpool-contextmanager-5981-2|

Some parts of `threadpool.py` would be clearer with context managers. For example, in `_worker`, `with working(self): ...` instead of matched append/remove calls.

Attachments:

* [5981_0.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/25e/25efb1446ad25211786ababd686d3bf51ce78783/ecb52e1c37a43f69cc6529aa710d047c23af761e.patch) (4269 bytes) - added by ephess on 2012-12-20 21:10:05Z - 
* [5981_1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/25e/25efb1446ad25211786ababd686d3bf51ce78783/7d24b8de5fabb7dfe2c1f428b862dc0a7b07c034.patch) (4121 bytes) - added by ephess on 2012-12-20 23:06:45Z - Patch addresses issues raised by review.
* [5981v3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/25e/25efb1446ad25211786ababd686d3bf51ce78783/25ba0e29668ec8a8019d38c13836c4029664a5ad.patch) (2303 bytes) - added by stephsolis on 2013-04-18 20:29:52Z - Addressed more issues from review and added news file

<details><summary>Searchable metadata</summary>

```
trac-id__5981 5981
type__enhancement enhancement
reporter__itamar itamar
priority__low low
milestone__ 
branch__branches_threadpool_contextmanager_5981_2 branches/threadpool-contextmanager-5981-2
branch_author__exarkun__glyph exarkun, glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1347632053000000 1347632053000000
changetime__1474198569965514 1474198569965514
version__None None
owner__ 
cc__ephess
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ephess's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|ephess commented|
|-|-|

Patch attached for sections of the _worker method. Have reviewed the rest of this module and can't find any other areas which look like they'd be more readable with a context manager - let me know if I missed anything.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [36612]) Branching to 'threadpool-contextmanager-5981'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [36613]) Apply 5981_0.patch 

refs [#5981](https://github.com/twisted/twisted/issues/5981)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Thanks!  This looks like a great patch.  I have just a few minor comments:

  1. Some names added in the patch don't follow the naming convention - we use `camelCase` rather than `under_scores` for variable names (amongst other things).
  1. A direct unit test or two for `_workerState` would be great, particularly for the exception case, which I don't think is currently exercised by the unit tests.
  1. This version involves a few extra calls to `currentThread`.  That probably doesn't matter, but since it's not too hard to avoid the calls (ie, make it an argument to `_workerState`), it might be nice.

I checked the patch into a branch.  If you can make the next patch against that branch rather than against trunk, that'd be great.  Thanks again!

(cannot reassign properly, _ephess_ is not in the user list for some reason :/)

**Comment by twisted-trac** [comment]
|<img alt="ephess's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ephess set owner to ephess<br>ephess set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="ephess's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|ephess removed owner<br>ephess set status to `new`|
|-|-|

Thanks for the feedback.

1. Fixed, hard habit to break!
2. Added, nice catch - there was a bug there.
3. Refactored to pass thread ID into context manager.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [36753]) Apply new patch

Refs: [#5981](https://github.com/twisted/twisted/issues/5981)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Hey have some [build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/threadpool-contextmanager-5981)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [36959]) Branching to 'threadpool-contextmanager-5981-2'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

[Re-building on new branch](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/threadpool-contextmanager-5981-2).

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_11"></a>|@tomprince set owner to ephess|
|-|-|

This looks good. A couple of minor comments:

  * `test_workerState` should probably be split into two tests, one for the non-exception case, and one for the exception case.
  * Also, it would be a good thing to check that `_workerState` doesn't change the exception that got raised (`1/0` and `ZeroDivisionError` are often used in twisted test-suite for this).

**Comment by twisted-trac** [comment]
|<img alt="stephsolis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stephsolis set owner to stephsolis<br>stephsolis set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="stephsolis's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|stephsolis removed owner<br>stephsolis set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_14"></a>|@therve commented|
|-|-|

(In [38442]) Apply patch

Refs [#5981](https://github.com/twisted/twisted/issues/5981)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_15"></a>|@therve set owner to @therve|
|-|-|

Build results: buildbot.twistedmatrix.com/boxes-supported?branch=/branches/threadpool-contextmanager-5981-2

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

Review, un-assigning.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_18"></a>|@therve set status to `closed`|
|-|-|

(In [38584]) Merge threadpool-contextmanager-5981-2

Authors: stephsolis, ephess
Reviewers: tom.prince, exarkun, therve
Fixes: [#5981](https://github.com/twisted/twisted/issues/5981)

Cleanup threadpool implementation by using context managers.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_19"></a>|@hawkowl commented|
|-|-|

[mass edit] Removing the review tag from closed tickets
